lib/OpenQA/Schema/ResultSet/Jobs.pm

```diff
@@ -84,13 +84,8 @@ MACHINE. For each set of dupes, only the latest job found is included in
 the return array.
 
 =cut
-sub latest_jobs ($self, $until = undef) { @{$self->latest_jobs_with_limit($until)->{jobs}} }
-
-sub latest_jobs_with_limit ($self, $until = undef, $limit = undef) {
-    my @jobs = $self->search(
-        $until ? {t_created => {'<=' => $until}} : undef,
-        {order_by => ['me.id DESC'], $limit ? (rows => $limit) : ()});
-    my $limit_exceeded = $limit ? @jobs >= $limit : undef;
+sub latest_jobs ($self, $until = undef) {
+    my @jobs = $self->search($until ? {t_created => {'<=' => $until}} : undef, {order_by => ['me.id DESC']});
```

Comment:
Your reasoning makes sense to me and as long as nobody can come up with a way to pass filter+limit parameters directly as according SQL commands and parameters I think the general approach is ok. I just wonder if we can still apply *some* reasonable limit when we actually query the database. Not necessarily to guard against malicious attacks by users but to protect ourselves against implementation bugs as well to prevent unlimited SQL queries, e.g. here, especially as you directly assign to an array so I assume the actual table rows are converted into the Perl array here already.

Reply:
Any limit applied here would be arbitrary and bound to cause similar issues. Moving the de-duplication to SQL is the only real solution imo. Fortunately even the worst case examples i came up with on my local O3 clone with many thousands of jobs never took longer than 2 seconds to hit the new 500 job limit in the Perl code.

Reply:
The real performance bottleneck here has always been the prefetching of comments. Which is pretty well shielded in this version. Everything else only has a minor impact (< 2s).

Reply:
You would fetch all of them, `DBD::Pg` has no support for paging, so it will always fetch all results at once. Here's a sample query generated by `DBIx::Class` for Tumbleweed jobs (82991 results, 1.3s to retrieve all):
```
EXECUTE SELECT me.id, me.result_dir, me.archived, me.state, me.priority, me.result, me.reason, me.clone_id, me.blocked_by_id, me.backend_info, me.TEST, me.DISTRI, me.VERSION, me.FLAVOR, me.ARCH, me.BUILD, me.MACHINE, me.group_id, me.assigned_worker_id, me.t_started, me.t_finished, me.logs_present, me.passed_module_count, me.failed_module_count, me.softfailed_module_count, me.skipped_module_count, me.externally_skipped_module_count, me.scheduled_product_id, me.result_size, me.t_created, me.t_updated FROM jobs me WHERE ( ( me.clone_id IS NULL AND me.VERSION = $1 ) ) ORDER BY me.id DESC (
$1: Tumbleweed
);
```

Reply:
So we do still have some headroom, but sooner or later we will have to deal with this and untangle the query generator.

Reply:
Hm, hoping for an opinion by other field experts.

Reply:
That would be good, it's a complex issue. Given the commit history of the affected functions, @coolo and @Martchus probably know the code and its history best.